Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,743 @@
+    PseudoEMetricSpace (WithLp p (α × β)) :=
+  (prodPseudoEmetricAux p α β).replaceUniformity (prod_aux_uniformity_eq p α β).symm
+
+/-- `EMetricSpace` instance on the product of two emetric spaces, using the `L^p`
+edistance, and having as uniformity the product uniformity. -/
+instance instProdEMetricSpace [EMetricSpace α] [EMetricSpace β] : EMetricSpace (WithLp p (α × β)) :=
+  @EMetricSpace.ofT0PseudoEMetricSpace (WithLp p (α × β)) _ instT0SpaceProdInstTopologicalSpaceProd
+
+/-- `PseudoMetricSpace` instance on the product of two pseudometric spaces, using the
+`L^p` distance, and having as uniformity the product uniformity. -/
+instance instProdPseudoMetricSpace [PseudoMetricSpace α] [PseudoMetricSpace β] :
+    PseudoMetricSpace (WithLp p (α × β)) :=
+  ((prodPseudoMetricAux p α β).replaceUniformity
+    (prod_aux_uniformity_eq p α β).symm).replaceBornology
+    fun s => Filter.ext_iff.1 (prod_aux_cobounded_eq p α β).symm sᶜ
+
+/-- `MetricSpace` instance on the product of two metric spaces, using the `L^p` distance,
+and having as uniformity the product uniformity. -/
+instance instProdMetricSpace [MetricSpace α] [MetricSpace β] : MetricSpace (WithLp p (α × β)) :=
+  MetricSpace.ofT0PseudoMetricSpace _
+
+variable {p α β}
+
+theorem prod_nndist_eq_add [PseudoMetricSpace α] [PseudoMetricSpace β]
+    (hp : p ≠ ∞) (x y : WithLp p (α × β)) :
+    nndist x y = (nndist x.fst y.fst ^ p.toReal + nndist x.snd y.snd ^ p.toReal) ^ (1 / p.toReal) :=
+  -- Porting note: was `Subtype.ext`
+  NNReal.eq <| by
+    push_cast
+    exact prod_dist_eq_add (p.toReal_pos_iff_ne_top.mpr hp) _ _
+
+theorem prod_nndist_eq_sup [PseudoMetricSpace α] [PseudoMetricSpace β] (x y : WithLp ∞ (α × β)) :
+    nndist x y = nndist x.fst y.fst ⊔ nndist x.snd y.snd :=
+  -- Porting note: was `Subtype.ext`
+  NNReal.eq <| by
+    push_cast
+    exact prod_dist_eq_sup _ _
+
+variable (p α β)
+
+theorem prod_lipschitzWith_equiv [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    LipschitzWith 1 (WithLp.equiv p (α × β)) :=
+  prod_lipschitzWith_equiv_aux p α β
+
+theorem prod_antilipschitzWith_equiv [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    AntilipschitzWith ((2 : ℝ≥0) ^ (1 / p).toReal) (WithLp.equiv p (α × β)) :=
+  prod_antilipschitzWith_equiv_aux p α β
+
+theorem infty_equiv_isometry [PseudoEMetricSpace α] [PseudoEMetricSpace β] :
+    Isometry (WithLp.equiv ∞ (α × β)) :=
+  fun x y =>
+  le_antisymm (by simpa only [ENNReal.coe_one, one_mul] using prod_lipschitzWith_equiv ∞ α β x y)
+    (by
+      simpa only [ENNReal.div_top, ENNReal.zero_toReal, NNReal.rpow_zero, ENNReal.coe_one,
+        one_mul] using prod_antilipschitzWith_equiv ∞ α β x y)
+
+/-- Seminormed group instance on the product of two normed groups, using the `L^p`
+norm. -/
+instance instProdSeminormedAddCommGroup [SeminormedAddCommGroup α] [SeminormedAddCommGroup β] :
+    SeminormedAddCommGroup (WithLp p (α × β)) :=
+  { WithLp.instAddCommGroup _ _ with
+    dist_eq := fun x y => by
```

Comment:
Can't you just write:
```suggestion
    SeminormedAddCommGroup (WithLp p (α × β)) where
  dist_eq := fun x y => by
```
and delete the trailing `}`? (And then adjust the indentation too)